server/asset/eth/rpcclient.go

```diff
@@ -33,61 +34,67 @@ type ContextCaller interface {
 	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
 }
 
-type rpcclient struct {
-	net      dex.Network
-	endpoint string
-	// ec wraps a *rpc.Client with some useful calls.
-	ec *ethclient.Client
-	// caller is a client for raw calls not implemented by *ethclient.Client.
-	caller ContextCaller
+type ethConn struct {
+	*ethclient.Client
 	// swapContract is the current ETH swapContract.
 	swapContract swapContract
-
 	// tokens are tokeners for loaded tokens. tokens is not protected by a
 	// mutex, as it is expected that the caller will connect and place calls to
 	// loadToken sequentially in the same thread during initialization.
 	tokens map[uint32]*tokener
-
+	// caller is a client for raw calls not implemented by *ethclient.Client.
+	caller          ContextCaller
 	txPoolSupported bool
 }
 
-func newRPCClient(net dex.Network, endpoint string) *rpcclient {
+type rpcclient struct {
+	net       dex.Network
+	log       dex.Logger
+	endpoints []string
+	clients   []*ethConn
+
+	idxMtx      sync.RWMutex
+	endpointIdx int
+}
+
+func newRPCClient(net dex.Network, endpoints []string, log dex.Logger) *rpcclient {
 	return &rpcclient{
-		net:      net,
-		tokens:   make(map[uint32]*tokener),
-		endpoint: endpoint,
+		net:       net,
+		endpoints: endpoints,
+		log:       log,
 	}
 }
 
-// connect connects to an ipc socket. It then wraps ethclient's client and
-// bundles commands in a form we can easily use.
-func (c *rpcclient) connect(ctx context.Context, log dex.Logger) error {
-	var client *rpc.Client
-	var err error
-	if strings.HasSuffix(c.endpoint, ".ipc") {
-		client, err = rpc.DialIPC(ctx, c.endpoint)
-		if err != nil {
-			return fmt.Errorf("unable to dial ipc: %v", err)
+func (c *rpcclient) withClient(f func(ec *ethConn) error, propagateNotFound ...bool) (err error) {
+	for range c.endpoints {
+		c.idxMtx.RLock()
+		idx := c.endpointIdx
+		ec := c.clients[idx]
+		c.idxMtx.RUnlock()
+
+		err = f(ec)
+		if err == nil {
+			return nil
 		}
-	} else {
-		log.Debugf("dialing endpoint: %s", c.endpoint)
-		client, err = rpc.DialContext(ctx, c.endpoint)
-		if err != nil {
-			return fmt.Errorf("unable to dial rpc: %v", err)
+		if len(propagateNotFound) > 0 && propagateNotFound[0] && strings.Contains(err.Error(), "not found") {
+			return err
```

Comment:
Digesting this variadic bool, it's more like `haltOnNotFound`, I think, like it short-circuits if it hits a not found from provider A so it doesn't bother with provider B because it's not really an error, just an expected outcome.